File: sdks/typescript/src/http.ts
import type { TokenStore } from "./auth.js";
import { PlynthApiError, PlynthNetworkError, parseErrorResponse } from "./errors.js";
import type { Tokens } from "./types.js";

const API_PREFIX = "/api/v1";

export interface HttpConfig {
  baseUrl: string;
  productSlug?: string;
  adminToken?: string;
  actingTenantSlug?: string;
  tokenStore: TokenStore;
  fetch?: typeof fetch;
  uuid?: () => string;
}

export interface RequestOptions {
  method: string;
  path: string;
  body?: unknown;
  query?: Record<string, string | number | boolean | undefined>;
  productSlug?: string;
  actingTenantSlug?: string;
  asPlatformAdmin?: boolean;
  skipAuth?: boolean;
  idempotent?: boolean;
  idempotencyKey?: string;
  signal?: AbortSignal;
}

function isAdminPath(path: string): boolean {
  return path.startsWith(`${API_PREFIX}/admin/`);
}

function defaultUuid(): string {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID();
  }
  // RFC4122 v4 fallback
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

export class HttpClient {
  readonly #cfg: HttpConfig;
  readonly #fetch: typeof fetch;
  readonly #uuid: () => string;

  constructor(cfg: HttpConfig) {
    this.#cfg = { ...cfg, baseUrl: cfg.baseUrl.replace(/\/+$/, "") };
    this.#fetch = cfg.fetch ?? globalThis.fetch.bind(globalThis);
    this.#uuid = cfg.uuid ?? defaultUuid;
  }

  async request<T>(opts: RequestOptions): Promise<T> {
    return this.#send<T>(opts, false);
  }

  async #send<T>(opts: RequestOptions, retried: boolean): Promise<T> {
    const headers = await this.#buildHeaders(opts);
    const url = this.#buildUrl(opts.path, opts.query);

    let res: Response;
    try {
      res = await this.#fetch(url, {
        method: opts.method,
        headers,
        body: opts.body !== undefined ? JSON.stringify(opts.body) : undefined,
        signal: opts.signal,
      });
    } catch (err) {
      throw new PlynthNetworkError(
        err instanceof Error ? err.message : String(err),
        err,
      );
    }

    const isUserCall = !opts.skipAuth && !opts.asPlatformAdmin && !isAdminPath(opts.path);
    if (res.status === 401 && isUserCall && !retried) {
      const tokens = await this.#cfg.tokenStore.get();
      if (tokens && (await this.#refresh(tokens))) {
        return this.#send<T>(opts, true);
      }
    }

    if (res.status === 204) return undefined as T;
    if (!res.ok) throw await parseErrorResponse(res);

    const text = await res.text();
    if (!text) return undefined as T;
    return JSON.parse(text) as T;
  }

  async #refresh(current: Tokens): Promise<boolean> {
    let res: Response;
    try {
      res = await this.#fetch(`${this.#cfg.baseUrl}${API_PREFIX}/auth/refresh`, {
        method: "POST",
        headers: { "Content-Type": "application/json", Accept: "application/json" },
        body: JSON.stringify({ refresh_token: current.refresh_token }),
      });
    } catch {
      await this.#cfg.tokenStore.clear();
      return false;
    }
    if (!res.ok) {
      await this.#cfg.tokenStore.clear();
      return false;
    }
    const next = (await res.json()) as Tokens;
    await this.#cfg.tokenStore.set(next);
    return true;
  }

  async #buildHeaders(opts: RequestOptions): Promise<Record<string, string>> {
    const headers: Record<string, string> = {
      Accept: "application/json",
    };
    if (opts.body !== undefined) headers["Content-Type"] = "application/json";

    const wantAdmin = opts.asPlatformAdmin === true || isAdminPath(opts.path);

    if (wantAdmin) {
      if (!this.#cfg.adminToken) {
        throw new PlynthApiError(401, {
          code: "no_platform_admin_token",
          message: "Platform admin token not configured on client.",
          details: {},
        });
      }
      headers["X-Platform-Admin-Token"] = this.#cfg.adminToken;
    } else if (!opts.skipAuth) {
      const tokens = await this.#cfg.tokenStore.get();
      if (tokens) {
        headers["Authorization"] = `Bearer ${tokens.access_token}`;
      } else if (this.#cfg.adminToken && this.#cfg.productSlug) {
        // Admin god-mode: no session, admin token + default product configured.
        headers["X-Platform-Admin-Token"] = this.#cfg.adminToken;
      }
    }

    const slug = opts.productSlug ?? this.#cfg.productSlug;
    if (slug) headers["X-Product-Slug"] = slug;

    const acting = opts.actingTenantSlug ?? this.#cfg.actingTenantSlug;
    if (acting && !wantAdmin) headers["X-Acting-Tenant-Slug"] = acting;

    if (opts.idempotent || opts.idempotencyKey) {
      headers["Idempotency-Key"] = opts.idempotencyKey ?? this.#uuid();
    }

    return headers;
  }

  #buildUrl(path: string, query?: RequestOptions["query"]): string {
    const url = new URL(`${this.#cfg.baseUrl}${path}`);
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
      }
    }
    return url.toString();
  }
}

File: sdks/python/plynth_sdk/_http.py
"""Shared header-build / refresh logic for both sync and async clients."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any

import httpx

from plynth_sdk.auth import TokenStore
from plynth_sdk.errors import PlynthApiError

API_PREFIX = "/api/v1"


def is_admin_path(path: str) -> bool:
    return path.startswith(f"{API_PREFIX}/admin/")


@dataclass
class HttpConfig:
    base_url: str
    token_store: TokenStore
    product_slug: str | None = None
    admin_token: str | None = None
    # Per-product service token (`pst_…`) for the product-runtime
    # `/env` path. When `RequestSpec.as_service_token=True` AND this
    # field is set, the SDK sends `X-Service-Token` and skips
    # Authorization / X-Platform-Admin-Token. Never expose to clients.
    service_token: str | None = None
    acting_tenant_slug: str | None = None
    timeout: httpx.Timeout = field(
        default_factory=lambda: httpx.Timeout(30.0, connect=10.0)
    )


@dataclass
class RequestSpec:
    method: str
    path: str
    json_body: Any | None = None
    params: dict[str, Any] | None = None
    product_slug: str | None = None
    acting_tenant_slug: str | None = None
    as_platform_admin: bool = False
    as_service_token: bool = False
    skip_auth: bool = False
    idempotent: bool = False
    idempotency_key: str | None = None


def build_headers(cfg: HttpConfig, spec: RequestSpec) -> dict[str, str]:
    headers: dict[str, str] = {"Accept": "application/json"}
    if spec.json_body is not None:
        headers["Content-Type"] = "application/json"

    want_admin = spec.as_platform_admin or is_admin_path(spec.path)
    want_service_token = spec.as_service_token

    if want_service_token:
        if not cfg.service_token:
            raise PlynthApiError(
                401,
                "no_service_token",
                "Service token not configured on client.",
            )
        headers["X-Service-Token"] = cfg.service_token
    elif want_admin:
        if not cfg.admin_token:
            raise PlynthApiError(
                401,
                "no_platform_admin_token",
                "Platform admin token not configured on client.",
            )
        headers["X-Platform-Admin-Token"] = cfg.admin_token
    elif not spec.skip_auth:
        tokens = cfg.token_store.get()
        if tokens:
            headers["Authorization"] = f"Bearer {tokens['access_token']}"
        elif cfg.admin_token and cfg.product_slug:
            # Admin god-mode: no user session + admin token configured.
            headers["X-Platform-Admin-Token"] = cfg.admin_token

    slug = spec.product_slug or cfg.product_slug
    if slug:
        headers["X-Product-Slug"] = slug

    acting = spec.acting_tenant_slug or cfg.acting_tenant_slug
    if acting and not want_admin:
        headers["X-Acting-Tenant-Slug"] = acting

    if spec.idempotent or spec.idempotency_key:
        headers["Idempotency-Key"] = spec.idempotency_key or str(uuid.uuid4())

    return headers


def parse_response(response: httpx.Response) -> Any:
    """Translate an httpx response into a parsed body or raise PlynthApiError."""
    from plynth_sdk.errors import parse_error_response

    if response.status_code == 204 or not response.content:
        if response.is_success:
            return None
        raise parse_error_response(response)
    try:
        body = response.json()
    except ValueError:
        body = None
    if not response.is_success:
        if isinstance(body, dict):
            raise PlynthApiError(
                response.status_code,
                str(body.get("code") or "http_error"),
                str(body.get("message") or response.reason_phrase or ""),
                body.get("details") if isinstance(body.get("details"), dict) else None,
            )
        raise PlynthApiError(
            response.status_code, "http_error", response.reason_phrase or "", None
        )
    return body
